### PR TITLE
Async Calls tests with multi transfer

### PR DESCRIPTION
**Pull request by dorin-iancu**
Added 2 tests for async call and callback with multi transfers
Fixed an edge case when the scr with "ok" message was deleted. AsyncCallback SCRs have to be created only if they are cross shard.

Test:
https://testnet-explorer.elrond.com/transactions/bf3b0c6b92992bb4a6a4a9e81e9ec0bf0e406dae92cded8451db9dd50a0d2a1c
Do this transaction and will observ 2 SCRs and will be treated as successful by indexer.

### CODE REVIEW

File: integrationTests/vm/txsFee/utils/utils.go [hunk 1]
```diff
@@ -195,6 +196,33 @@ func TestAccount(
 	return senderRecovShardAccount.GetBalance()
 }
 
+func CreateSmartContractCall(
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/utils/utils.go:199]: missing comment.

File: integrationTests/vm/txsFee/utils/utilsESDT.go [hunk 2]
```diff
@@ -60,10 +71,36 @@ func CreateAccountWithESDTBalanceAndRoles(
...
+	esdtNonce uint64,
 	esdtValue *big.Int,
 	roles [][]byte,
 ) {
-	CreateAccountWithESDTBalance(t, accnts, pubKey, egldValue, tokenIdentifier, esdtValue)
+	CreateAccountWithESDTBalance(t, accnts, pubKey, egldValue, tokenIdentifier, esdtNonce, esdtValue)
+	SetESDTRoles(t, accnts, pubKey, tokenIdentifier, roles)
+}
+
+func SetESDTRoles(
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/utils/utilsESDT.go:82]: missing comments on 3 new functions

File: process/smartContract/process.go [hunk 3]
```diff
@@ -1929,68 +1924,104 @@ func (sc *scProcessor) createAsyncCallBackSCRFromVMOutput(
...
-		if result.Value.Cmp(zero) > 0 {
-			result.OriginalSender = tx.GetSndAddr()
-		}
+		result = sc.preprocessOutTransferToSCR(i, outputTransfer, outAcc, tx, txHash)
 
+		isCrossShard := sc.shardCoordinator.ComputeId(outAcc.Address) != sc.shardCoordinator.SelfId()
 		if result.CallType == vmData.AsynchronousCallBack {
-			createdAsyncCallBack = true
-			result.GasLimit, _ = core.SafeAddUint64(result.GasLimit, vmOutput.GasRemaining)
+			if !sc.flagCreatedCallBackCrossShardOnly.IsSet() || isCrossShard {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:2020]: originally, the L2021-L2023 were not conditioned by being only cross-shard. Is this correct for backwards compatibility reasons? What happened if we had an intra shard async call?

File: process/smartContract/process.go [hunk 4]
```diff
@@ -2004,14 +2035,16 @@ func (sc *scProcessor) createSmartContractResults(
...
-			sc.useLastTransferAsAsyncCallBackWhenNeeded(callType, outAcc, &outputTransferCopy, vmOutput, tx, result) {
+			sc.useLastTransferAsAsyncCallBackWhenNeeded(callType, outAcc, &outputTransferCopy, vmOutput, tx, result, isCrossShard) {
 			createdAsyncCallBack = true
 		}
 
 		if result.CallType == vmData.AsynchronousCall {
-			result.GasLimit += outputTransfer.GasLocked
-			lastArgAsGasLocked := "@" + hex.EncodeToString(big.NewInt(0).SetUint64(outputTransfer.GasLocked).Bytes())
-			result.Data = append(result.Data, []byte(lastArgAsGasLocked)...)
+			if !sc.flagCreatedCallBackCrossShardOnly.IsSet() || isCrossShard {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:2043]: also here, we weren't conditioned by being cross-shard in the old version

File: process/smartContract/process.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/process.go:2020]: if we had intra shard async call - the gasLimit would be added to this SCR and the SCR for sender with @ok would not be created.

File: process/smartContract/process.go (hunk 4 above)
Comment (sasurobert) [reply process/smartContract/process.go:2043]: the gasLocked would be added and encoded to the SCR which was not needed, as intra shard async calls and callbacks were already resolved by arwen.

File: process/smartContract/process.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/process.go:2020]: intra shard async calls and callbacks were already resolved by arwen. this is needed for cross shard only.

File: integrationTests/vm/txsFee/utils/utils.go (hunk 1 above)
Comment (sasurobert) [reply integrationTests/vm/txsFee/utils/utils.go:199]: done

File: integrationTests/vm/txsFee/utils/utilsESDT.go (hunk 2 above)
Comment (sasurobert) [reply integrationTests/vm/txsFee/utils/utilsESDT.go:83]: done

File: process/smartContract/process.go (hunk 3 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:2020]: ok

**Review by iulianpascalau — DISMISSED** [review]

File: integrationTests/vm/testInitializer.go [hunk 5]
```diff
@@ -783,32 +786,35 @@ func CreateTxProcessorWithOneSCExecutorWithVMs(
...
-		ScrForwarder:                        intermediateTxHandler,
-		BadTxForwarder:                      intermediateTxHandler,
-		TxFeeHandler:                        feeAccumulator,
-		EconomicsFee:                        economicsData,
-		TxTypeHandler:                       txTypeHandler,
-		GasHandler:                          gasComp,
-		GasSchedule:                         mock.NewGasScheduleNotifierMock(gasSchedule),
-		TxLogsProcessor:                     &mock.TxLogsProcessorStub{},
-		EpochNotifier:                       forking.NewGenericEpochNotifier(),
+	enableEpochs := config.EnableEpochs{
```
Comment (bogdan-rosianu) [inline integrationTests/vm/testInitializer.go:789]: this approach is dangerous as we won't have any feedback if the value wasn't set and this way we'll get default values (0). in the future, we might think to have a separate component that handles enable epochs and it's functions return `(uint32, error)`.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed. 
image

**Review by bogdan-rosianu — APPROVED** [review]

File: genesis/process/metaGenesisBlockCreator.go [hunk 6]
```diff
@@ -381,8 +381,8 @@ func createProcessorsForMetaGenesisBlock(arg ArgsGenesisBlockCreator, enableEpoc
 		ScProcessor:                           scProcessor,
 		TxTypeHandler:                         txTypeHandler,
 		EconomicsFee:                          genesisFeeHandler,
-		ESDTEnableEpoch:                       arg.EpochConfig.EnableEpochs.ESDTEnableEpoch,
-		BuiltInFunctionOnMetachainEnableEpoch: arg.EpochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
+		ESDTEnableEpoch:                       enableEpochs.ESDTEnableEpoch,
```
Comment (iulianpascalau) [inline genesis/process/metaGenesisBlockCreator.go:384]: we should follow @bogdan-rosianu suggestion an refactor these config passing using a specialized struct. This will be a new PR

**Review by iulianpascalau — APPROVED** [review]
